documentation/guide/writing-persistent-objects.rst

```diff
@@ -0,0 +1,533 @@
+   >>> conn = db.open()
+   >>> book = conn.root.book
+   >>> bool(book._p_changed)
+   False
+   >>> book.authors.append('Jim')
+   >>> bool(book._p_changed)
+   False
+   >>> book.add_author('Carlos')
+   >>> bool(book._p_changed)
+   True
+   >>> db.close()
+
+In this example, ``Book`` objects have an ``authors`` object that's a
+regular Python list, so it's mutable and non-persistent.  When we add
+an author, we append it to the ``authors`` attribute's value.  Because
+we didn't set an attribute on the book, it's not marked as changed, so
+we set ``_p_changed`` ourselves.
+
+Using standard Python lists, dicts, or sets is a common thing to do,
+so this pattern of setting ``_p_changed`` is common.
+
+Let's look at some alternatives.
+
+Using tuples for small sequences instead of lists
+------------------------------------------------
+
+If objects contain sequences that are small or that don't change
+often, you can use tuples instead of lists::
+
+  import persistent
+
+  class Book(persistent.Persistent):
+
+     def __init__(self, title):
+         self.title = title
+         self.authors = ()
+
+     def add_author(self, author):
+         self.authors += (author, )
+
+.. -> src
+
+   >>> exec(src)
+   >>> db = ZODB.DB(None)
+   >>> with db.transaction() as conn:
+   ...     conn.root.book = Book("ZODB")
+   >>> conn = db.open()
+   >>> book = conn.root.book
+   >>> bool(book._p_changed)
+   False
+   >>> book.add_author('Carlos')
+   >>> bool(book._p_changed)
+   True
+   >>> db.close()
+
+Because tuples are immutable, they satisfy the rules of persistence
+without any special handling.
+
+Using persistent data structures
+--------------------------------
+
+The ``persistent`` package provides persistent versions of ``list``
+and ``dict``, namely ``persistent.list.PersistentList`` and
+``persistent.mapping.PersistentMapping``. We can update our example to
+use ``PersistentList``::
+
+  import persistent
+  import persistent.list
+
+  class Book(persistent.Persistent):
+
+     def __init__(self, title):
+         self.title = title
+         self.authors = persistent.list.PersistentList()
+
+     def add_author(self, author):
+         self.authors.append(author)
+
+.. -> src
+
+   >>> exec(src)
+   >>> db = ZODB.DB(None)
+   >>> with db.transaction() as conn:
+   ...     conn.root.book = Book("ZODB")
+   >>> conn = db.open()
+   >>> book = conn.root.book
+   >>> bool(book._p_changed)
+   False
+   >>> book.add_author('Carlos')
+   >>> bool(book._p_changed)
+   False
+   >>> bool(book.authors._p_changed)
+   True
+   >>> db.close()
+
+Note that in this example, when we added an author, the book itself
+didn't change, but the ``authors`` attribute value did.  Because
+``authors`` is a persistent object, it's stored in a separate database
+record from the book record and is managed by ZODB independent of the
+management of the book.
+
+In addition to ``PersistentList`` and ``PersistentMapping``, general
```

Comment:
We used to use it, but even with it resolving conflicts, our (pretty high-volume write) use case turned out to be better served by Redis. But it works for what it does.